### ISSUE TITLE
Support for visitor/doSwitch-like feature

### ISSUE THREAD

**Issue by aranega** [pyecore/pyecore#23]
Following a [discussion](https://twitter.com/tarelli/status/984066583067848704) on twitter with @tarelli, it would indeed be great that PyEcore proposes a solution that is the equivalent of the `doSwitch`.

I give it a though, and it seems that the logic behind the `doSwitch` is almost equivalent to the `@singledispatch` decorator from the `functools` lib (and there is a backport for Python 2.7). To match the behavior, I could be able to produce a decorator, e.g: `@object_visitor` (whatever the name), that could be used like this (in a metamodel with `A`, `B`, and `C` metaclasses):

```python
class MyVisitor(object):
  @object_visitor
  def fun(self, arg):
    print('fallback for', arg)

  @fun.register(A)
  def caseA(self, arg):
    print('In case for A', arg)

  @fun.register(B)
  def caseB(self, arg):
    print('in case for B', arg)
```

And you would use it a little bit like the one in EMF (let say we have a `root`):

```python
visitor = MyVisistor()
for obj in root.eAllContents():
  visitor.fun(obj)
```

This solution could also be used as base for a `ComposedSwitch` like the one in EMF. Currently, the advantages I see are that it relies on a standard library, and it would work in a same way for dynamic and static metamodels (which is great), However, a drawback is that you have to add the default `fun` function yourself and cannot inherit it from another component and should have the same name (not obligatory, but will help for `ComposedSwitch`, this could be enforce anyway).

@tarelli Would a `doSwitch` functionnality like this do the job?

**Comment by tarelli** [comment]
@aranega yes, that would be great! Thanks so much Vincent, next time you are in London give me a shout, drinks on me :)

**Comment by nicain** [comment]
This would be straight up magic!

**Comment by aranega** [comment]
@tarelli @nicain Thanks! Be sure I keep this in mind next time I go to London or first time I will be in Seattle ;)

I performed some experiment, and it was surprisingly easy to implement as Python already almost have everything we need, and it works in a transparent way for dynamic and static metamodels (which is great if you enhance your metamodel at runtime). However, the behavior I have differs a little bit from the EMF one:

* In EMF, each 'Switch' is dedicated to an `EPackage`, meaning that if your metamodel is built with many inner `EPackage`, you need to use a `ComposedSwitch` with a `Switch` for each inner `EPackage`. With the experiment I've done, you can use the same 'Switch' for pretty much any metaclass you want, everything relies on the way you will structured and design your code. It also means that `ComposedSwitch` are pretty much useless (from what I understand).
* In EMF, when no element from a metamodel is found, the `defaultCase()` is called, and when an element from the metamodel is called, but there is no `caseXXX()` implemented, nothing happend. With the experiment, if no method matches the element class, the main `do_switch()`, which embeds the default code, is called.
* In EMF, the return value of an `caseXXX()` method controls the fact that the default method should be called or not. This is not the case with the exepriment, if you also want the default case, you must call it. For this one, I don't like the EMF way, I understand how it could be handy, but using a return value to control such a thing feels a little odd. 

This experiment uses the `singledispatch` as basis and tweaks it a little bit. I can write something from scratch that will enforce the exact same behavior than EMF, but I'm not sure that replicating at 100% the original behavior is mandatory. 

Would you be ok with the differences I mentionned? That would add a new dependency for Python 2.7 only as `singledispatch` already exists in Python >= 3.4.
I feel that the behavior I described would be great for anyone, but I perhaps forgot about some corner cases.

**Comment by aranega** [comment]
A preview of the `dispatch` decorator is on the `develop` branch of PyEcore and PyEcore-py2. It still misses the documentation, but you can use it like this (for a metamodel containing `A` and `B` metaclasses):

```python
from pyecore.dispatch import dispatch

# Switch definition
class MySwitch(object):
    @dispatch
    def do_switch(self, o):
        print('In default case')

    @do_switch.register(A)
    def caseA(self, o):
        print('Case', o.eClass.name)

    @do_switch.register(B)
    def caseB(self, o):
        print(Case for B instances', o) 


# let say that we have a 'root', and we want to iterate on every children
# 'eAllContent()' produces a generator
switcher = MySwitch()
for child in root.eAllContents():
    switcher.do_switch(child)
````

Currently, there is no restriction on the default function name. This would have interesting for `ComposedSwitch`, but as the switch does not bring the same constraints has EMF, it seems that a `ComposedSwitch` would not be necessary .

**Comment by aranega** [comment]
I had time today to make some modifications on the code. I moved the `@dispatch` feature from `pyecore.dispatch` to `pyecore.utils`, I also added a dedicated entry in the [documentation](http://pyecore.readthedocs.io/en/latest/user/quickstart.html#traversing-the-whole-model-with-single-dispatch), and a new version of PyEcore is available!

If the feature sounds good to you, I will close the issue.
